File: packages/server/src/shared/rate-limit.ts
import type { FastifyRequest, FastifyReply } from 'fastify';
import { redis } from './redis.js';
import { AppError } from './errors.js';

class RateLimitError extends AppError {
  constructor(retryAfterSec: number, message?: string) {
    super('RATE_LIMITED', message ?? `Too many requests. Retry after ${retryAfterSec} seconds.`, 429);
    this.retryAfter = retryAfterSec;
  }
  retryAfter: number;
}

interface RateLimitOptions {
  /** Redis key prefix */
  prefix: string;
  /** Max requests allowed in the window */
  max: number;
  /** Window size in seconds */
  windowSec: number;
  /** Extract the key to rate-limit by (IP, email, API key hash, etc.) */
  keyFrom: (request: FastifyRequest) => string | null;
  /** Custom message when rate limited */
  message?: string;
}

/**
 * Redis sliding-window rate limiter using INCR + EXPIRE.
 * Returns a Fastify preHandler hook.
 */
export function rateLimit(opts: RateLimitOptions) {
  return async (request: FastifyRequest, reply: FastifyReply): Promise<void> => {
    const id = opts.keyFrom(request);
    if (!id) return; // can't identify → skip

    const key = `flick:rl:${opts.prefix}:${id}`;

    const count = await redis.incr(key);
    if (count === 1) {
      // First request in this window — set TTL
      await redis.expire(key, opts.windowSec);
    }

    const ttl = await redis.ttl(key);
    reply.header('X-RateLimit-Limit', opts.max);
    reply.header('X-RateLimit-Remaining', Math.max(0, opts.max - count));
    reply.header('X-RateLimit-Reset', Math.ceil(Date.now() / 1000) + Math.max(ttl, 0));

    if (count > opts.max) {
      const retryAfter = Math.max(ttl, 1);
      reply.header('Retry-After', retryAfter);
      const msg = opts.message
        ? `${opts.message} Retry after ${retryAfter}s.`
        : `Too many requests. Retry after ${retryAfter} seconds.`;
      throw new RateLimitError(retryAfter, msg);
    }
  };
}

// ── Pre-built limiters ──

/** Login: 5 attempts per IP per 60s */
export const loginRateLimit = rateLimit({
  prefix: 'login',
  max: 5,
  windowSec: 60,
  keyFrom: (req) => req.ip,
  message: 'Whoa, easy there! Too many login attempts. Go grab a coffee and try again.',
});

/** Login by email: 10 attempts per email per 5min (cross-IP brute-force) */
export const loginEmailRateLimit = rateLimit({
  prefix: 'login-email',
  max: 10,
  windowSec: 300,
  keyFrom: (req) => {
    const body = req.body as { email?: string } | undefined;
    return body?.email?.toLowerCase() ?? null;
  },
  message: "That's a lot of password guesses. Maybe try the 'Forgot password' button?",
});

/** SDK evaluate: 1000 requests per API key per 10s (6000 rpm) */
export const evaluateRateLimit = rateLimit({
  prefix: 'eval',
  max: 1000,
  windowSec: 10,
  keyFrom: (req) => {
    const auth = req.headers.authorization;
    if (!auth?.startsWith('Bearer ')) return null;
    // Use the last 8 chars as identifier — no need to hash again
    return auth.slice(-8);
  },
  message: "Dude, you're not supposed to call /evaluate on every request. Use the SDK — it polls and caches flags locally. Read the docs!",
});

File: packages/server-rust/src/middleware/rate_limit.rs
use axum::extract::Request;
use axum::http::HeaderValue;
use axum::middleware::Next;
use axum::response::Response;
use fred::prelude::*;

use crate::app::AppState;
use crate::cache::keys;
use crate::error::AppError;

pub struct RateLimitConfig {
    pub prefix: &'static str,
    pub max: i64,
    pub window_sec: i64,
    pub message: &'static str,
}

pub const SDK_EVALUATE_LIMIT: RateLimitConfig = RateLimitConfig {
    prefix: "eval",
    max: 1000,
    window_sec: 10,
    message: "Too many evaluation requests. Use the SDK — it polls and caches flags locally.",
};

pub const LOGIN_IP_LIMIT: RateLimitConfig = RateLimitConfig {
    prefix: "login",
    max: 5,
    window_sec: 60,
    message: "Too many login attempts.",
};

pub const LOGIN_EMAIL_LIMIT: RateLimitConfig = RateLimitConfig {
    prefix: "login-email",
    max: 10,
    window_sec: 300,
    message: "Too many login attempts for this email.",
};

/// Check rate limit using Redis INCR + EXPIRE sliding window.
pub async fn check_rate_limit(
    redis: &fred::clients::RedisPool,
    config: &RateLimitConfig,
    identifier: &str,
) -> Result<RateLimitHeaders, AppError> {
    let key = keys::rate_limit_key(config.prefix, identifier);

    let count: i64 = redis.incr(&key).await.map_err(|e| {
        tracing::warn!(error = %e, "Rate limit INCR failed");
        AppError::Internal(anyhow::anyhow!("Rate limit check failed"))
    })?;

    if count == 1 {
        let _ = redis.expire::<(), _>(&key, config.window_sec).await;
    }

    let ttl: i64 = redis.ttl(&key).await.unwrap_or(config.window_sec);
    let remaining = (config.max - count).max(0);
    let reset = chrono::Utc::now().timestamp() + ttl.max(0);

    if count > config.max {
        let retry_after = ttl.max(1);
        return Err(AppError::rate_limited(
            format!("{} Retry after {}s.", config.message, retry_after),
            retry_after,
        ));
    }

    Ok(RateLimitHeaders {
        limit: config.max,
        remaining,
        reset,
    })
}

pub struct RateLimitHeaders {
    pub limit: i64,
    pub remaining: i64,
    pub reset: i64,
}

impl RateLimitHeaders {
    pub fn apply(&self, response: &mut Response) {
        let headers = response.headers_mut();
        headers.insert(
            "X-RateLimit-Limit",
            HeaderValue::from_str(&self.limit.to_string()).unwrap(),
        );
        headers.insert(
            "X-RateLimit-Remaining",
            HeaderValue::from_str(&self.remaining.to_string()).unwrap(),
        );
        headers.insert(
            "X-RateLimit-Reset",
            HeaderValue::from_str(&self.reset.to_string()).unwrap(),
        );
    }
}

/// Middleware for SDK evaluate rate limiting (keyed by last 8 chars of bearer token).
pub async fn evaluate_rate_limit(
    axum::extract::State(state): axum::extract::State<AppState>,
    request: Request,
    next: Next,
) -> Result<Response, AppError> {
    let identifier = request
        .headers()
        .get("authorization")
        .and_then(|v| v.to_str().ok())
        .filter(|v| v.starts_with("Bearer "))
        .map(|v| &v[v.len().saturating_sub(8)..]);

    if let Some(id) = identifier {
        let headers = check_rate_limit(&state.redis, &SDK_EVALUATE_LIMIT, id).await?;
        let mut response = next.run(request).await;
        headers.apply(&mut response);
        Ok(response)
    } else {
        Ok(next.run(request).await)
    }
}
